src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionManager.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.ServiceHub.Client;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
+    {
+        private partial class ConnectionManager
+        {
+            private readonly HubClient _hubClient;
+            private readonly HostGroup _hostGroup;
+            private readonly TimeSpan _timeout;
+
+            private readonly ReaderWriterLockSlim _shutdownLock;
+            private readonly ReferenceCountedDisposable<RemotableDataJsonRpc> _remotableDataRpc;
+
+            private readonly int _maxPoolConnections;
+
+            // keyed to serviceName. each connection is for specific service such as CodeAnalysisService
+            private readonly ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>> _pools;
+
+            // indicate whether pool should be used.
+            private readonly bool _enableConnectionPool;
+
+            public ConnectionManager(
+                HubClient hubClient,
+                HostGroup hostGroup,
+                bool enableConnectionPool,
+                int maxPoolConnection,
+                TimeSpan timeout,
+                ReferenceCountedDisposable<RemotableDataJsonRpc> remotableDataRpc)
+            {
+                _hubClient = hubClient;
+                _hostGroup = hostGroup;
+                _timeout = timeout;
+
+                _remotableDataRpc = remotableDataRpc;
+                _maxPoolConnections = maxPoolConnection;
+
+                // we have 4 services. so start from 4. later if we add more services, it will still work.
```

Comment:
Can we not hard code this number?

Reply:
not sure what you mean. we want to give some initial value to it otherwise, concurrentdictionary goes crazy. you can see how we use concurrentdictionary in other places.